x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/saml/SamlCompleteLogoutRequest.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.core.security.action.saml;
+
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+
+import java.io.IOException;
+import java.util.List;
+
+import static org.elasticsearch.action.ValidateActions.addValidationError;
+
+/**
+ * Represents a request to complete SAML LogoutResponse
+ */
+public final class SamlCompleteLogoutRequest extends ActionRequest {
+
+    @Nullable
+    private String queryString;
+    @Nullable
+    private String content;
+    private List<String> validRequestIds;
+    private String realm;
+
+    public SamlCompleteLogoutRequest(StreamInput in) throws IOException {
+        super(in);
+    }
+
+    public SamlCompleteLogoutRequest() {
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException validationException = null;
+        if (Strings.hasText(realm) == false) {
+            validationException = addValidationError("realm may not be empty", validationException);
+        }
+        if (Strings.hasText(queryString) == false && Strings.hasText(content) == false) {
+            validationException = addValidationError("queryString and content may not both be empty", validationException);
+        }
+        if (Strings.hasText(queryString) && Strings.hasText(content)) {
+            validationException = addValidationError("queryString and content may not both present", validationException);
```

Comment:
What is our expectation for how a WebApp should behave if it gets a LogoutResponse over a HTTP-POST binding that also has URL parameters?

I think the rule for Kibana is:
- If the http method is `GET` send the queryString to Elasticsearch, and ignore (or reject) any body (which is very unlikely to exist)
- if the http method is `POST` ignore (or reject) the query parameters, and send the body to Elasticsearch.

Is that our intent?

Reply:
A good question. To be honest, I haven't really considered this thoroughly. I think your suggestion makes sense because the code proceeds as HTTP-POST, e.g. base64 encode, no deflate and XML signature, when `content` is present. This basically says `content` implies `POST`, which in turn means `queryString` implies `GET`.

I would personally prefer to ignore the body for `GET` and query string for `POST` since the spec does not explicitly forbid them.